Use Thor for built-in stats task

JuanVqz opened the pull request:
### Motivation / Background

Currently, we use both Thor and Rake for `bin/rails` commands. We eventually want to get all the built-ins task promoted to Thor Commands. This migrates the `stats` task to Thor.

### Detail



### Additional information



### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

JuanVqz committed:
Use Thor for built-in stats task

Currently, we use both Thor and Rake for `bin/rails` commands.
We eventually want to get all the built-ins task promoted to Thor Commands.
This migrates the `stats` task to Thor.

p8 commented on railties/lib/rails/commands/statistics/statistics_command.rb:
```diff
@@ -26,12 +26,21 @@
+        require "rails/code_statistics"
+
+        stat_directories = STATS_DIRECTORIES.collect do |name, dir|
+          [name, "#{File.dirname(Rake.application.rakefile_location)}/#{dir}"]
```
I think we can now do the following:
```suggestion
          [name, Rails::Command.application_root.join(dir)]
```
Similar to https://github.com/rails/rails/blob/a51364dcf1dab34bc6cfc9aeca56a3b0ecf823ad/railties/lib/rails/commands/restart/restart_command.rb#L9

JuanVqz commented on railties/lib/rails/commands/statistics/statistics_command.rb:
```diff
@@ -26,12 +26,21 @@
+    class StatisticsCommand < Base # :nodoc:
+      desc :stats, "Report code statistics (KLOCs, etc) from the application or engine"
+      def perform
+        require "rails/code_statistics"
```
previously we used the following piece of code in the `railties/lib/rails/tasks.rb` 

```
tap { |arr|
  arr << "statistics" if Rake.application.current_scope.empty?
}
```
Do you think this is a good equivalent of what it did?

```suggestion
        return if Rake.application.current_scope.any?
        
        require "rails/code_statistics"
```

p8 replied:
I don't think that is still a problem when it's no longer a Rake task.
If you look at the commit that added it, it seems related to engines loading the task twice.
As it is now a Thor task I don't think it will be loaded twice.
https://github.com/rails/rails/commit/38f28f77049e7160491b71167471a2175253b742

JuanVqz replied:
That makes sense, thank you!

p8 commented:
It looks like `statistics.rake` is still included in the Rakefile of engines/plugins:
https://github.com/rails/rails/blob/main/railties/lib/rails/generators/rails/plugin/templates/Rakefile.tt#L9
Removing it would break a lot of plugins.
We probably need keep the file but show a deprecation warning instead.

JuanVqz commented on railties/lib/rails/tasks/statistics.rake:
```diff
@@ -33,5 +33,6 @@ end.select { |name, dir| File.directory?(dir) }
 desc "Report code statistics (KLOCs, etc) from the application or engine"
 task :stats do
   require "rails/code_statistics"
+  ActiveSupport::Deprecation.warn("STATS_DIRECTORIES will be moved to an initializer file")
```
not pretty sure if this is the right deprecation warning message :)

p8 commented:
@JuanVqz https://github.com/rails/rails/pull/47313 got merged into main.
So you should probably rebase on main.

JuanVqz commented:
@p8 Hello again!

I updated the PR with the changes from https://github.com/rails/rails/pull/47313. I added a deprecation warning and the new Rails command for the `stats`. Please review it and let me know if I need to update something else, possibly the changelog.

Thanks!

zzak commented:
@JuanVqz Would you mind sending a PR to update the [changelog](https://github.com/rails/rails/blob/main/railties/CHANGELOG.md)? :bow:

p8 commented:
@zzak @JuanVqz I haven't add Juan to my changelog entry and I'm not sure it will get merged.
So I think we'll need a changelog entry for the rake task deprecation.

JuanVqz commented:
@zzak thanks for the heads up! I'll add a note in the changelog next time

JuanVqz commented:
@p8 @zzak didn't see p8 comment I'll open a pr

JuanVqz commented:
@p8 @zzak is this note good enough? https://github.com/rails/rails/pull/52231

willnet commented on railties/lib/rails/tasks/statistics.rake:
```diff
@@ -34,5 +34,11 @@ task :stats do
   end.select { |name, dir| File.directory?(dir) }
+
+  $stderr.puts Rails.deprecator.warn(<<~MSG, caller_locations(0..1))
+  `bin/rails stats` as rake task has been deprecated and will be removed in Rails 8.0.
```
This deprecation warning will be visible starting with the release of Rails 8.0, so it seems correct to use “in Rails 8.1” instead of “in Rails 8.0”.
